src/drivers/camera_capture/camera_capture.cpp

```diff
@@ -0,0 +1,313 @@
+		struct camera_trigger_s trigger = {};
+		_trigger_pub = orb_advertise(ORB_ID(camera_trigger), &trigger);
+	}
+}
+
+CameraCapture::~CameraCapture()
+{
+	camera_capture::g_camera_capture = nullptr;
+}
+
+void
+CameraCapture::capture_callback(uint32_t chan_index,
+				hrt_abstime edge_time, uint32_t edge_state, uint32_t overflow)
+{
+
+	if (_last_fall_time > 0) {
+		struct camera_trigger_s	trigger {};
+
+		if (_camera_capture_mode == 0) {
+			trigger.timestamp = edge_time;
+
+		} else {
+			trigger.timestamp = edge_time - ((edge_time - _last_fall_time) / 2);	// Get timestamp of mid-exposure
+		}
+
+		trigger.seq = _capture_seq++;
+
+		if (_camera_capture_feedback) {
+			orb_publish(ORB_ID(camera_trigger), _trigger_pub, &trigger);
+		}
+
+		_last_exposure_time = edge_time - _last_fall_time;
+	}
+
+	// Timestamp and compensate for strobe delay
+	_last_fall_time = edge_time - uint64_t(1000 * _strobe_delay);
+
+	_capture_overflows = overflow;
+
+}
+
+void
+CameraCapture::capture_trampoline(void *context, uint32_t chan_index,
+				  hrt_abstime edge_time, uint32_t edge_state, uint32_t overflow)
+{
+	camera_capture::g_camera_capture->capture_callback(chan_index, edge_time, edge_state, overflow);
+}
+
+void
+CameraCapture::cycle_trampoline(void *arg)
+{
+
+	CameraCapture *cap = reinterpret_cast<CameraCapture *>(arg);
+
+	if (cap->_command_sub < 0) {
+		cap->_command_sub = orb_subscribe(ORB_ID(vehicle_command));
+	}
+
+	bool updated = false;
+	orb_check(cap->_command_sub, &updated);
+
+	// Command handling
+	if (updated) {
+
+		vehicle_command_s cmd;
+		orb_copy(ORB_ID(vehicle_command), cap->_command_sub, &cmd);
+
+		// TODO : this should eventuallly be a capture control command
+		if (cmd.command == vehicle_command_s::VEHICLE_CMD_DO_TRIGGER_CONTROL) {
+
+			// Enable/disable signal capture
+			if (commandParamToInt(cmd.param1) == 1) {
+				cap->set_capture_control(true);
+
+			} else if (commandParamToInt(cmd.param1) == 0) {
+				cap->set_capture_control(false);
+
+			}
+
+			// Reset capture sequence
+			if (commandParamToInt(cmd.param2) == 1) {
+				cap->reset_statistics(true);
+
+			}
+
+			// Acknowledge the command
+			vehicle_command_ack_s command_ack = {
+				.timestamp = 0,
+				.result_param2 = 0,
+				.command = cmd.command,
+				.result = (uint8_t)vehicle_command_s::VEHICLE_CMD_RESULT_ACCEPTED,
+				.from_external = false,
+				.result_param1 = 0,
+				.target_system = cmd.source_system,
+				.target_component = cmd.source_component
+			};
+
+			if (cap->_command_ack_pub == nullptr) {
+				cap->_command_ack_pub = orb_advertise_queue(ORB_ID(vehicle_command_ack), &command_ack,
+							vehicle_command_ack_s::ORB_QUEUE_LENGTH);
+
+			} else {
+				orb_publish(ORB_ID(vehicle_command_ack), cap->_command_ack_pub, &command_ack);
+
+			}
+		}
+
+	}
+
+	work_queue(LPWORK, &_work, (worker_t)&CameraCapture::cycle_trampoline, camera_capture::g_camera_capture,
+		   USEC2TICK(100000)); // 100ms
+}
+
+void
+CameraCapture::set_capture_control(bool enabled)
```

Comment:
The coordination is done using a mode command to the fmu. as seen here https://github.com/PX4/Firmware/blob/master/src/drivers/px4fmu/fmu.cpp#L3048-L3050 

You will need to expand the fmu to support V5 captures.

Then use the IOCTL https://github.com/PX4/Firmware/blob/master/src/drivers/px4fmu/fmu.cpp#L2314

Here is an example https://github.com/PX4/Firmware/blob/master/src/drivers/px4fmu/fmu.cpp#L2625-L2650

Reply:
Isn't this an opportunity to move away from the fmu dependency entirely?

Reply:
It could be. But one would have to move all the code that manages the pwm and capture and all the pins to another module. There needs to be a single point of control and configuration. As it stands it was the FMU

Reply:
Fwiw, FMU input capture is currently broken and crashes the board if the IOCTL is used. I filed an issue sometime back.

Reply:
@mhkabir  It was fixed at one point perhaps not upstreamed. Did you test on current master?

Reply:
I tried this quickly today but was not working. This branch is about a week behind master. @davids5 where can I see if this was fixed?